### ISSUE TITLE
Remove cmp usage in twisted.names.srvconnect

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5124]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| facundobatista reported|
|-|-|
|Trac ID|trac#5124|
|Type|enhancement|
|Created|2011-05-28 13:26:38Z|
|Branch|https://github.com/twisted/twisted/tree/names-cmp-5124|

The attached patch removes `cmp` and `__cmp__` usage from the following modules:

 - [ticket:6059 twisted/names/dns.py]
 - `twisted/names/srvconnect.py`

All the changes are tested (had to add some tests for some changes).

Attachments:

* [remove-cmp-02a.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f88/f88269a2d0ad75d2ed6a16e785dc7eb334799276/9dc157fc40eeabb8a4c4fdf19701dd4a398afc41.patch) (9076 bytes) - added by facundobatista on 2011-05-28 13:27:23Z - 
* [remove-cmp-02b.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f88/f88269a2d0ad75d2ed6a16e785dc7eb334799276/ed197bbab6063a528977662cd515a05d6c69ce38.patch) (8126 bytes) - added by facundobatista on 2011-06-01 02:42:19Z - Changes in 'names' only
* [remove-cmp-02c.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f88/f88269a2d0ad75d2ed6a16e785dc7eb334799276/e742a39c17c2669d1d41c8a12c2676b990ab483a.patch) (8129 bytes) - added by facundobatista on 2011-06-14 13:11:06Z - Removed the assertLess calls.
* [remove-cmp-02d.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f88/f88269a2d0ad75d2ed6a16e785dc7eb334799276/1d5f5a21ee9f7f5af9b9389aee4a48f4a1cae11d.patch) (9690 bytes) - added by facundobatista on 2011-06-30 01:01:45Z - Support all comparisons, with more tests

<details><summary>Searchable metadata</summary>

```
trac-id__5124 5124
type__enhancement enhancement
reporter__facundobatista facundobatista
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_names_cmp_5124 branches/names-cmp-5124
branch_author__facundobatista__thijs facundobatista, thijs
status__closed closed
resolution__fixed fixed
component__names names
keywords__py3k py3k
time__1306589198000000 1306589198000000
changetime__1465913311552501 1465913311552501
version__None None
owner__facundobatista facundobatista
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

fixup ticket description.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra set owner to facundobatista|
|-|-|

Thanks for the patch.

`compareMarkPos` and 'comparePosition' in [trunk/twisted/lore/tree.py] are deprecated since Twisted 9.0, so they shouldn't be updated but removed instead. I guess that needs a new ticket because every removal needs a ticket. So I suggest opening a new one for lore with a patch that removes those deprecated methods, and ignoring the lore changes in the patch on this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Don't spent time on Lore maintenance.  Work on the Sphinx conversion. :)

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|facundobatista commented|
|-|-|

thijs: I created ticket [#5127](https://github.com/twisted/twisted/issues/5127) for the removal of these methods, and attached a new patch here with the changes in `names` only. Thanks!

exarkun: I don't want to spend any time in Lore, but I need those `cmp` out of the way for Py3k :).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

Replying to [facundobatista](#note_4):
> thijs: I created ticket [#5127](https://github.com/twisted/twisted/issues/5127) for the removal of these methods, and attached a new patch here with the changes in `names` only. Thanks!

Thanks, I'll update the ticket title and summary to reflect this update.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

(In [32112]) Branching to 'names-cmp-5124'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

(In [32113]) apply remove-cmp-02b.patch, add news file. refs [#5124](https://github.com/twisted/twisted/issues/5124)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

(In [32114]) pyflakes and coding std fixes, refs [#5124](https://github.com/twisted/twisted/issues/5124)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra set owner to facundobatista|
|-|-|

The [build results](http://buildbot.twistedmatrix.com/builders/lucid64-py2.6-select/builds/1004) show 3 failing tests. Can you make new patches against this branch?

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|facundobatista commented|
|-|-|

Attached the 'c' version of the patch, where I removed the assertLess calls (that were only 2.7, sorry).

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|facundobatista removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to facundobatista|
|-|-|

It appears that this patch has a glaring omission: to wit, tests for `>` comparison.  Is there a rule about `__gt__` that it automatically works if `__lt__`, `__eq__` and `__ne__` are all implemented?  In any case I think more test coverage is required...

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|facundobatista removed owner|
|-|-|

Yes, __gt__ automatically works if __lt__ is implemented and both are the same type (because of reflection when doing the comparison... Python doesn't find __gt__ on one object, tries __lt__ on the other).

However, the <= comparisons weren't working... I added code for that, and tests cases for everything. 

'c' version of the patch is attached.

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|facundobatista commented|
|-|-|

Bah, sorry, the version of the patch for these last improvements is the 'd' one.

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_17"></a>|@dustin set owner to facundobatista|
|-|-|

In twisted/names/dns.py, the fallback for __le__ uses < instead of <=.  The effect is negligible - a Query is never equal to anything else, so <= and < are equivalent, but the difference stands out in the code and looks suspicious, so it should be corrected.  On that note, what is the point of ```self.__class__ == other.__class__``` if you already know that ```other``` is not an instance of ```Query```?  Why not just return False from ```__eq___``` in that case?

I can't see any other problems - all of the comments in the previous reviews seem to be addressed.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra commented|
|-|-|

[#5348](https://github.com/twisted/twisted/issues/5348) was marked as a duplicate.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_19"></a>|@thijstriemstra commented|
|-|-|

The first part of this ticket (twisted.names.dns) was fixed with [#6059](https://github.com/twisted/twisted/issues/6059).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_20"></a>|@ralphm set status to `closed`|
|-|-|

In changeset 5aee182293607ad4b806d81c26dbbd1f6e1f4624
```
#!CommitTicketReference repository="" revision="5aee182293607ad4b806d81c26dbbd1f6e1f4624"
Merge py3-srvconnect-8262: Port twisted.names.srvconnect to Python 3.

Author: ralphm
Reviewer: adiroiban, hawkowl, glyph
Fixes: #8262, #5124, #2800

This also cleans up the use of a cmp function for sorting SRV records,
makes the `pickServer` implementation in line with the algorithm
specified in the RFC and adds tests around that to match.
```
